### PR TITLE
raftstore-v2: implement CaptureChange and LeaderCallback

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: ref #14542

What's Changed:

```commit-message
raftstore-v2: implement CaptureChange and LeaderCallback
```

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/router/imp.rs [hunk 1]
```diff
@@ -179,19 +180,72 @@ impl<EK: KvEngine, ER: RaftEngine> RaftRouter<EK, ER> {
...
 
     fn check_leadership(
         &self,
-        _region_id: u64,
-        _callback: Callback<EK::Snapshot>,
+        region_id: u64,
+        callback: Callback<EK::Snapshot>,
     ) -> crate::Result<()> {
-        unimplemented!()
+        let (ch, _) = QueryResChannel::with_callback(Box::new(|res| {
```
Comment (YuJuncen) [inline components/raftstore-v2/src/router/imp.rs:227]: Should we move `callback` in?

File: components/raftstore-v2/src/operation/query/capture.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        // TODO: add a test case for this kind of situation.
+
+        let apply_router = self.fsm.peer().apply_scheduler().unwrap().clone();
+        let (ch, _) = QueryResChannel::with_callback(Box::new(move |res| {
+            if let QueryResult::Response(resp) = res && resp.get_header().has_error() {
+                // Return error
+                capture_change.snap_cb.report_error(resp.clone());
+                return;
+            }
+            apply_router.send(ApplyTask::CaptureApply(capture_change))
```
Comment (YuJuncen) [inline components/raftstore-v2/src/operation/query/capture.rs:54]: A (maybe dummy) question: if we asynchronously change the capture state, is it possible to lose data between `[read_index_idx, change_capture_state_idx)`? Or `apply_router` has a synchronous nature?

File: components/raftstore-v2/src/raft/apply.rs [hunk 3]
```diff
@@ -59,6 +62,8 @@ pub struct Apply<EK: KvEngine, R> {
     res_reporter: R,
     read_scheduler: Scheduler<ReadTask<EK>>,
     sst_importer: Arc<SstImporter>,
+    observe_info: CmdObserveInfo,
```
Comment (YuJuncen) [inline components/raftstore-v2/src/raft/apply.rs:65]: (I'm also not pretty sure about how this influences to the apply. Will the logic of observing be sent in another PR?)

**Review by YuJuncen — COMMENTED** [review]
Generally LGTM, with some little questions.

File: components/raftstore-v2/src/raft/apply.rs (hunk 3 above)
Comment (overvenus) [reply components/raftstore-v2/src/raft/apply.rs:65]: Yes, see https://github.com/tikv/tikv/pull/14562

File: components/raftstore-v2/src/operation/query/capture.rs (hunk 2 above)
Comment (overvenus) [reply components/raftstore-v2/src/operation/query/capture.rs:54]: The read index here is fow 1) ensure it's a leader and 2) the its data is ready to read. The data between `[read_index_idx, change_capture_state_idx)` is included in snapshot which is taken by `Apply`.

**Review by YuJuncen — APPROVED** [review]

File: components/raftstore-v2/src/operation/query/capture.rs (hunk 2 above)
Comment (YuJuncen) [reply components/raftstore-v2/src/operation/query/capture.rs:54]: I see, we get the snapshot when handling the `CaptureChange` instead of after `ReadIndex`.

File: components/raftstore-v2/src/router/imp.rs (hunk 1 above)
Comment (YuJuncen) [reply components/raftstore-v2/src/router/imp.rs:227]: Wow, amazing... 😮

File: components/raftstore-v2/src/router/response_channel.rs [hunk 4]
```diff
@@ -471,6 +472,36 @@ impl CmdResChannelBuilder {
     }
 }
 
+pub type AnyResChannel = BaseChannel<(RaftCmdResponse, Option<Box<dyn Any + Send + Sync>>)>;
```
Comment (3pointer) [inline components/raftstore-v2/src/router/response_channel.rs:475]: How about expand CmdResChannel instead of create new channel?

**Review by 3pointer — APPROVED** [review]

File: components/raftstore-v2/src/router/response_channel.rs (hunk 4 above)
Comment (overvenus) [reply components/raftstore-v2/src/router/response_channel.rs:475]: CmdResChannel does not accept `Box<dyn Any + Send + Sync>`.

**Review by Connor1996 — APPROVED** [review]
